tools/make_patch.c

```diff
@@ -0,0 +1,437 @@
+			fprintf(output, "0x%0*x%03X", padding - 3, value >> 12, value & 0xfff);
+		} else {
+			fprintf(output, isupper((unsigned char)command[0]) ? "0x%0*X" : "0x%0*x", padding, value);
+		}
+
+	} else {
+		error_exit("Error: Unknown command: \"%s\"\n", command);
+	}
+}
+
+void skip_to_next_line(FILE *restrict input, FILE *restrict output) {
+	for (int c = getc(input); c != EOF; c = getc(input)) {
+		putc(c, output);
+		if (c == '\n' || c == '\r') {
+			break;
+		}
+	}
+}
+
+struct Buffer *process_template(const char *template_filename, const char *patch_filename, FILE *restrict new_rom, FILE *restrict orig_rom, const struct Symbol *symbols) {
+	FILE *input = xfopen(template_filename, 'r');
+	FILE *output = xfopen(patch_filename, 'w');
+	struct Buffer *buffer = buffer_create(1);
+
+	struct Buffer *patches = buffer_create(sizeof(struct Patch));
+
+	// The ROM checksum will always differ
+	buffer_append(patches, &(struct Patch){0x14e, 2});
+	// The Stadium data (see stadium.c) will always differ
+	unsigned int rom_size = (unsigned int)xfsize("", orig_rom);
+	unsigned int stadium_size = 24 + 6 + 2 + (rom_size / 0x2000) * 2;
+	buffer_append(patches, &(struct Patch){rom_size - stadium_size, stadium_size});
+
+	const struct Symbol *current_hook = NULL;
+
+	// Fill in the template
+	for (int c = getc(input); c != EOF; c = getc(input)) {
+		switch (c) {
+		case ';':
+			// ";" comments until the end of the line
+			putc(c, output);
+			skip_to_next_line(input, output);
+			break;
+
+		case '{':
+			// "{...}" is a template command; buffer its contents
+			buffer->size = 0;
+			for (c = getc(input); c != EOF && c != '}'; c = getc(input)) {
+				buffer_append(buffer, &c);
+			}
+			buffer_append(buffer, &(char []){'\0'});
+			// Interpret the command in the context of the current patch
+			interpret_command(buffer->data, current_hook, symbols, patches, new_rom, orig_rom, output);
+			break;
+
+		case '[':
+			// "[...]" is a patch label; buffer its contents
+			putc(c, output);
+			buffer->size = 0;
+			for (c = getc(input); c != EOF; c = getc(input)) {
+				putc(c, output);
+				if (c == ']') {
+					break;
+				} else if (!isalnum(c) && c != '_' && c != '@' && c != '#') {
+					// Convert non-identifier characters to underscores
+					c = '_';
+				}
+				buffer_append(buffer, &c);
+			}
+			buffer_append(buffer, &(char []){'\0'});
+			// The current patch should have a corresponding ".VC_" label
+			current_hook = symbol_find_cat(symbols, ".VC_", buffer->data);
+			skip_to_next_line(input, output);
+			break;
+
+		default:
+			putc(c, output);
+		}
+	}
+
+	rewind(orig_rom);
+	rewind(new_rom);
+
+	fclose(input);
+	fclose(output);
+	buffer_free(buffer);
+	return patches;
+}
+
+int compare_patch(const void *patch1, const void *patch2) {
+	unsigned int offset1 = ((const struct Patch *)patch1)->offset;
+	unsigned int offset2 = ((const struct Patch *)patch2)->offset;
+	return offset1 > offset2 ? 1 : offset1 < offset2 ? -1 : 0;
```

Comment:
Are you sure `return offset1 > offset2` isn't enough? That's what I do any time I use qsort.

Reply:
Yes, the `qsort` spec requires a "comparison function which returns ​a negative integer value if the first argument is less than the second, a positive integer value if the first argument is greater than the second and zero if the arguments are equivalent."

Reply:
Oh right, then just return `offset1 - offset2`.

Reply:
That works for the range of offsets we actually use, but in general, implementing `compare(a, b)` as `a - b` could cause signed overflow, which is undefined behavior. `return (offset1 > offset2) - (offset1 < offset2)` would also work, but I think the current form is the most obvious.